### PR TITLE
Django: capture responseHeadersOnEntrySpans

### PR DESCRIPTION
**Pull request by GSVarsha**
Django: capture responseHeadersOnEntrySpans

Signed-off-by: Varsha GS [varsha.gs@ibm.com](mailto:varsha.gs@ibm.com)

### CODE REVIEW

File: instana/instrumentation/django/middleware.py [hunk 1]
```diff
@@ -29,19 +29,27 @@ def __init__(self, get_response=None):
         super(InstanaMiddleware, self).__init__(get_response)
         self.get_response = get_response
 
+    def _extract_custom_headers(self, span, headers, format):
+        try:
+            if agent.options.extra_http_headers is not None:
```
Comment (Ferenc-) [inline instana/instrumentation/django/middleware.py:34]: This check should be inverted for an early return and should go touside of the `try` block.

File: tests/apps/app_django.py [hunk 2]
```diff
@@ -125,10 +125,17 @@ def complex(request):
     return HttpResponse('Stan wuz here!')
 
 
+def response_with_headers(request):
+    response = HttpResponse(content_type='')
+    response['X-Capture-This-Too'] = 'this too'
+    return response
```
Comment (Ferenc-) [inline tests/apps/app_django.py:131]:
```suggestion
    return HttpResponse('Stan wuz here with headers!', headers={'X-Capture-This-Too': 'this too'})
```

Comment (Ferenc-) [reply tests/apps/app_django.py:131]: I think it is still safer if we cover the capture of a list of headers, and not just a single header. Although the `test_request_header_capture` tests with a list and currently that covers the same codepath, but that is an implementational detail which at least theoretically could change in the future.

File: tests/frameworks/test_django.py [hunk 3]
```diff
@@ -306,6 +307,49 @@ def test_custom_header_capture(self):
         assert "X-Capture-That" in django_span.data["http"]["header"]
         self.assertEqual("that", django_span.data["http"]["header"]["X-Capture-That"])
 
+        agent.options.extra_http_headers = original_extra_http_headers
+
+    def test_response_header_capture(self):
+        # Hack together a manual custom headers list
+        original_extra_http_headers = agent.options.extra_http_headers
+        agent.options.extra_http_headers = [u'X-Capture-This-Too']
```
Comment (Ferenc-) [inline tests/frameworks/test_django.py:315]: I think it is safer to test with more than one headers at a time.

File: tests/frameworks/test_django.py [hunk 4]
```diff
@@ -306,6 +307,49 @@ def test_custom_header_capture(self):
...
+
+        with tracer.start_active_span('test'):
+            response = self.http.request('GET', self.live_server_url + '/response_with_headers')
+
+        assert response
```
Comment (Ferenc-) [inline tests/frameworks/test_django.py:320]: I know this was all over this TS, but this should really be `self.assertTrue`.

File: tests/frameworks/test_django.py [hunk 5]
```diff
@@ -306,6 +307,49 @@ def test_custom_header_capture(self):
...
+
+        self.assertEqual(None, django_span.ec)
+        self.assertIsNone(django_span.stack)
+
+        self.assertEqual('/response_with_headers', django_span.data["http"]["url"])
+        self.assertEqual('GET', django_span.data["http"]["method"])
+        self.assertEqual(200, django_span.data["http"]["status"])
+        self.assertEqual('^response_with_headers$', django_span.data["http"]["path_tpl"])
+
+        assert "X-Capture-This-Too" in django_span.data["http"]["header"]
```
Comment (Ferenc-) [inline tests/frameworks/test_django.py:348]:
```suggestion
        self.assertIn( "X-Capture-This-Too", django_span.data["http"]["header"])
```

File: tests/frameworks/test_django.py [hunk 6]
```diff
@@ -306,6 +307,49 @@ def test_custom_header_capture(self):
...
+        self.assertEqual("this too", django_span.data["http"]["header"]["X-Capture-This-Too"])
+
```
Comment (Ferenc-) [inline tests/frameworks/test_django.py:350]: If we had another header to capture, then we could validate it here.

**Review by Ferenc- — CHANGES_REQUESTED** [review]
Good job! I have suggested a few minor changes.

**Review by Ferenc- — APPROVED** [review]
LGTM

**Review by pvital — APPROVED** [review]
It looks good to me.

**Pull request merged**